File: android/app/src/main/java/jp/senri/arukou/StepCounterStore.java
package jp.senri.arukou;

import android.content.Context;
import android.content.SharedPreferences;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * 端末の TYPE_STEP_COUNTER（起動後累計）から日次歩数を永続化する。
 */
public final class StepCounterStore {

    private static final String PREFS = "senri_daily_steps";

    private static final String KEY_TRACKING = "tracking_enabled";
    private static final String KEY_DAY = "day_key";
    private static final String KEY_BASELINE = "day_baseline";
    private static final String KEY_LAST_COUNTER = "last_counter";
    private static final String KEY_TODAY = "today_steps";
    private static final String KEY_PENDING = "pending_days_json";
    private static final String KEY_UPDATED_AT = "updated_at_ms";

    private StepCounterStore() {}

    private static SharedPreferences prefs(Context context) {
        return context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public static String todayKey() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        return sdf.format(new Date());
    }

    public static boolean isTrackingEnabled(Context context) {
        return prefs(context).getBoolean(KEY_TRACKING, false);
    }

    public static void setTrackingEnabled(Context context, boolean enabled) {
        prefs(context).edit().putBoolean(KEY_TRACKING, enabled).apply();
    }

    public static synchronized void onCounterReading(Context context, float rawCounter) {
        SharedPreferences prefs = prefs(context);
        int counter = Math.max(0, (int) rawCounter);
        String day = todayKey();
        String savedDay = prefs.getString(KEY_DAY, "");
        int baseline = prefs.getInt(KEY_BASELINE, counter);
        int lastCounter = prefs.getInt(KEY_LAST_COUNTER, counter);

        if (counter < lastCounter) {
            baseline = counter;
            savedDay = day;
        }

        if (!day.equals(savedDay)) {
            int previousDaySteps = prefs.getInt(KEY_TODAY, 0);
            if (!savedDay.isEmpty() && previousDaySteps > 0) {
                addPendingDay(context, savedDay, previousDaySteps);
            }
            baseline = counter;
            savedDay = day;
        }

        int todaySteps = Math.max(0, counter - baseline);

        prefs.edit()
            .putString(KEY_DAY, savedDay)
            .putInt(KEY_BASELINE, baseline)
            .putInt(KEY_LAST_COUNTER, counter)
            .putInt(KEY_TODAY, todaySteps)
            .putLong(KEY_UPDATED_AT, System.currentTimeMillis())
            .commit();
    }

    public static int getTodaySteps(Context context) {
        checkAndRollDay(context);
        return Math.max(0, prefs(context).getInt(KEY_TODAY, 0));
    }

    public static long getUpdatedAtMs(Context context) {
        return prefs(context).getLong(KEY_UPDATED_AT, 0L);
    }

    /** 日付が変わったら本日歩数をリセットし、前日分を保留キューへ保存 */
    public static synchronized void checkAndRollDay(Context context) {
        SharedPreferences prefs = prefs(context);
        String day = todayKey();
        String savedDay = prefs.getString(KEY_DAY, "");
        if (savedDay.isEmpty()) {
            prefs.edit().putString(KEY_DAY, day).apply();
            return;
        }
        if (day.equals(savedDay)) {
            return;
        }

        int previousDaySteps = prefs.getInt(KEY_TODAY, 0);
        if (previousDaySteps > 0) {
            addPendingDay(context, savedDay, previousDaySteps);
        }

        int counter = prefs.getInt(KEY_LAST_COUNTER, 0);
        prefs.edit()
            .putString(KEY_DAY, day)
            .putInt(KEY_BASELINE, counter)
            .putInt(KEY_TODAY, 0)
            .apply();
    }

    public static void resetToday(Context context) {
        SharedPreferences prefs = prefs(context);
        int counter = prefs.getInt(KEY_LAST_COUNTER, 0);
        prefs.edit()
            .putString(KEY_DAY, todayKey())
            .putInt(KEY_BASELINE, counter)
            .putInt(KEY_TODAY, 0)
            .apply();
    }

    private static void addPendingDay(Context context, String dayKey, int steps) {
        try {
            SharedPreferences prefs = prefs(context);
            JSONArray arr = new JSONArray(prefs.getString(KEY_PENDING, "[]"));
            boolean updated = false;
            for (int i = 0; i < arr.length(); i++) {
                JSONObject item = arr.getJSONObject(i);
                if (dayKey.equals(item.optString("date"))) {
                    item.put("steps", Math.max(item.optInt("steps", 0), steps));
                    updated = true;
                    break;
                }
            }
            if (!updated) {
                JSONObject item = new JSONObject();
                item.put("date", dayKey);
                item.put("steps", steps);
                arr.put(item);
            }
            prefs.edit().putString(KEY_PENDING, arr.toString()).apply();
        } catch (Exception ignored) {
            /* ignore */
        }
    }

    public static List<JSONObject> getPendingDays(Context context) {
        List<JSONObject> out = new ArrayList<>();
        try {
            JSONArray arr = new JSONArray(prefs(context).getString(KEY_PENDING, "[]"));
            for (int i = 0; i < arr.length(); i++) {
                JSONObject item = arr.getJSONObject(i);
                if (item.optInt("steps", 0) > 0) {
                    out.add(item);
                }
            }
        } catch (Exception ignored) {
            /* ignore */
        }
        return out;
    }

    public static void clearPendingDays(Context context) {
        prefs(context).edit().putString(KEY_PENDING, "[]").apply();
    }

    public static void removePendingDay(Context context, String dayKey) {
        try {
            SharedPreferences prefs = prefs(context);
            JSONArray arr = new JSONArray(prefs.getString(KEY_PENDING, "[]"));
            JSONArray next = new JSONArray();
            for (int i = 0; i < arr.length(); i++) {
                JSONObject item = arr.getJSONObject(i);
                if (!dayKey.equals(item.optString("date"))) {
                    next.put(item);
                }
            }
            prefs.edit().putString(KEY_PENDING, next.toString()).apply();
        } catch (Exception ignored) {
            /* ignore */
        }
    }
}

File: ios/App/App/StepCounterStore.swift
import Foundation

enum StepCounterStore {
    private static let prefs = UserDefaults.standard
    private static let keyTracking = "senri_daily_steps.tracking_enabled"
    private static let keyDay = "senri_daily_steps.day_key"
    private static let keyToday = "senri_daily_steps.today_steps"
    private static let keyPending = "senri_daily_steps.pending_days_json"

    static func todayKey(from date: Date = Date()) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    static func startOfToday(from date: Date = Date()) -> Date {
        Calendar.current.startOfDay(for: date)
    }

    static func isTrackingEnabled() -> Bool {
        prefs.bool(forKey: keyTracking)
    }

    static func setTrackingEnabled(_ enabled: Bool) {
        prefs.set(enabled, forKey: keyTracking)
    }

    static func getTodaySteps() -> Int {
        rollDayIfNeeded()
        return max(0, prefs.integer(forKey: keyToday))
    }

    static func saveTodaySteps(_ steps: Int, date: Date = Date()) {
        let day = todayKey(from: date)
        let savedDay = prefs.string(forKey: keyDay) ?? ""

        if savedDay != day {
            let previousSteps = max(0, prefs.integer(forKey: keyToday))
            if !savedDay.isEmpty && previousSteps > 0 {
                addPendingDay(dateKey: savedDay, steps: previousSteps)
            }
            prefs.set(day, forKey: keyDay)
        }

        prefs.set(max(0, steps), forKey: keyToday)
    }

    static func resetToday() {
        prefs.set(todayKey(), forKey: keyDay)
        prefs.set(0, forKey: keyToday)
    }

    static func rollDayIfNeeded() {
        let day = todayKey()
        let savedDay = prefs.string(forKey: keyDay) ?? ""
        if savedDay.isEmpty {
            prefs.set(day, forKey: keyDay)
            return
        }
        if savedDay != day {
            let previousSteps = max(0, prefs.integer(forKey: keyToday))
            if previousSteps > 0 {
                addPendingDay(dateKey: savedDay, steps: previousSteps)
            }
            prefs.set(day, forKey: keyDay)
            prefs.set(0, forKey: keyToday)
        }
    }

    static func getPendingDays() -> [[String: Any]] {
        guard
            let raw = prefs.string(forKey: keyPending),
            let data = raw.data(using: .utf8),
            let json = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else {
            return []
        }
        return json.filter { ($0["steps"] as? Int ?? 0) > 0 }
    }

    static func clearPendingDays() {
        prefs.set("[]", forKey: keyPending)
    }

    static func removePendingDay(_ dateKey: String) {
        var items = getPendingDays()
        items.removeAll { ($0["date"] as? String) == dateKey }
        savePendingDays(items)
    }

    static func addPendingDayIfNeeded(dateKey: String, steps: Int) {
        addPendingDay(dateKey: dateKey, steps: steps)
    }

    static func dayInterval(for dateKey: String) -> (start: Date, end: Date)? {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "yyyy-MM-dd"
        guard let day = formatter.date(from: dateKey) else { return nil }
        let start = Calendar.current.startOfDay(for: day)
        guard let next = Calendar.current.date(byAdding: .day, value: 1, to: start) else { return nil }
        let end = min(Date(), next.addingTimeInterval(-0.001))
        if end <= start { return nil }
        return (start, end)
    }

    static func shiftDateKey(_ dateKey: String, days: Int) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "yyyy-MM-dd"
        guard let day = formatter.date(from: dateKey),
              let shifted = Calendar.current.date(byAdding: .day, value: days, to: day) else {
            return dateKey
        }
        return todayKey(from: shifted)
    }

    static func dateKeysInclusive(from: String, to: String) -> [String] {
        guard from <= to else { return [] }
        var keys: [String] = []
        var current = from
        var guardCount = 0
        while current <= to && guardCount < 400 {
            keys.append(current)
            current = shiftDateKey(current, days: 1)
            guardCount += 1
        }
        return keys
    }

    private static func addPendingDay(dateKey: String, steps: Int) {
        var items = getPendingDays()
        if let index = items.firstIndex(where: { ($0["date"] as? String) == dateKey }) {
            let current = items[index]["steps"] as? Int ?? 0
            items[index]["steps"] = max(current, steps)
        } else {
            items.append(["date": dateKey, "steps": steps])
        }
        savePendingDays(items)
    }

    private static func savePendingDays(_ items: [[String: Any]]) {
        guard let data = try? JSONSerialization.data(withJSONObject: items),
              let raw = String(data: data, encoding: .utf8) else {
            return
        }
        prefs.set(raw, forKey: keyPending)
    }
}
